### DISCUSSION TITLE
Interested in understanding of how AI-powered Subdomain Mutations work

### DISCUSSION THREAD

**Discussion by Greatz08** [blacklanternsecurity/bbot#1321] [General]
First of all i would like to thank you for this awesome open source project dev :-)).
I am interested in understanding about this feature which is mentioned in repo main page - "AI-powered Subdomain Mutations"
What it is and how it works and what is ai thing in this and overall working of it.
Thanks in advance x)

**Comment by TheTechromancer** [comment]
Thanks for your interest in the subdomain mutations. I should start by saying that our method for this is a very simple implementation of NLP, but it's both efficient and effective.

We don't use large language models or anything so sophisticated (although I plan to experiment with this in the future). Instead, over the course of a BBOT scan, every discovered DNS name is analyzed and converted into mutations, which are accumulated into a "mutator" that counts the occurrences of each one.

This analysis is key. We use the [wordninja](https://github.com/keredson/wordninja) library, trained on a DNS-specific dataset, to split the hostname as many ways as possible, including intelligent splitting by word. The code for this is [here](https://github.com/blacklanternsecurity/bbot/blob/8f358b66185b00869a0a2e662b471e823f5b17d1/bbot/core/helpers/wordcloud.py#L452).

Finally at the end of the scan, we take the most commonly occurring mutations and apply them recursively to all subdomains found by the scan.
